Plugins/Rename/Rename.cpp

```diff
@@ -0,0 +1,248 @@
+}
+
+Rename::~Rename()
+{
+}
+
+CNWSPlayer *Rename::player(Types::ObjectID playerId)
+{
+
+    if (playerId == Constants::OBJECT_INVALID)
+    {
+        LOG_NOTICE("NWNX_Rename function called on OBJECT_INVALID");
+        return nullptr;
+    }
+
+    auto *pPlayer = Globals::AppManager()->m_pServerExoApp->GetClientObjectByObjectId(playerId);
+    if (!pPlayer)
+    {
+        LOG_NOTICE("NWNX_Rename function called on non-player object %x", playerId);
+    }
+    return pPlayer;
+}
+
+void Rename::HookPlayerList(Services::Hooks::CallType cType, CNWSMessage* message, CNWSPlayer* pPlayer)
+{
+    Rename& plugin = *g_plugin;
+    if (cType == Services::Hooks::CallType::BEFORE_ORIGINAL)
+    {
+        //traverse the player list and replace the names of all players on the override list
+        plugin.GlobalNameChange(false);
+    }
+    else
+    { 
+        //And now we do it again to restore the names after the player list has been sent over 
+        plugin.GlobalNameChange(true);
+    }
+}
+
+void Rename::HookPartyInvite(Services::Hooks::CallType cType, NWNXLib::API::CNWSMessage* message, CNWSPlayer* pPlayer, unsigned char c)
+{
+    Rename& plugin = *g_plugin;
+    if (cType == Services::Hooks::CallType::BEFORE_ORIGINAL)
+    {
+        //traverse the player list and replace the names of all players on the override list
+        plugin.GlobalNameChange(false);
+    }
+    else
+    {
+        //And now we do it again to restore the names after the party invite
+        plugin.GlobalNameChange(true);
+    }
+}
+
+void Rename::GlobalNameChange(bool bOriginal)
+{
+    API::CServerExoAppInternal* server = Globals::AppManager()->m_pServerExoApp->m_pcExoAppInternal;
+ 
+    //go through all the players and change their names or restore them based on bOriginal.
+    
+    API::CExoLinkedListInternal* playerList = server->m_pNWSPlayerList->m_pcExoLinkedListInternal;
+    
+    for (API::CExoLinkedListNode* head = playerList->pHead; head; head = head->pNext) 
+    {
+        CNWSClient* client = reinterpret_cast<API::CNWSClient*>(head->pObject);
+        Types::ObjectID pcObjectID = static_cast<CNWSPlayer*>(client)->m_oidNWSObject;
+        std::string lsFirstName = (bOriginal) ?  *g_plugin->GetServices()->m_perObjectStorage->Get<std::string>(pcObjectID,firstNameKey) :  *g_plugin->GetServices()->m_perObjectStorage->Get<std::string>(pcObjectID,overrideNameKey);
+        std::string lsLastName = (bOriginal) ?  *g_plugin->GetServices()->m_perObjectStorage->Get<std::string>(pcObjectID,lastNameKey) : std::string("");
+        CNWSCreature* pCreature = server->GetCreatureByGameObjectID(pcObjectID);
+        
+        if (pCreature && !lsFirstName.empty() && ( bOriginal || !pCreature->m_sDisplayName.IsEmpty()))
+        {
+            pCreature->m_pStats->m_lsFirstName = ContainString(lsFirstName);
+            pCreature->m_pStats->m_lsLastName = ContainString(lsLastName);
+        }
+    }
+}
+
+std::string Rename::ExtractString(CExoLocString locStr)
```

Comment:
This will create a copy of the parameter, with the additional memory allocations and everything. A better option would be to have the argument as `const CExoLocString& locStr` to pass it by (constant) reference. Same below.